06_gpu_and_ml/reinforcement-learning/grpo-verl.py

```diff
@@ -0,0 +1,297 @@
+
+
+def extract_solution(
+    solution_str: str, method: Literal["strict", "flexible"] = "strict"
+) -> Optional[str]:
+    assert method in ["strict", "flexible"]
+
+    if method == "strict":
+        # This also tests the formatting of the model
+        solution = re.search("#### (\\-?[0-9\\.\\,]+)", solution_str)
+        if solution is None:
+            final_answer: Optional[str] = None
+        else:
+            final_answer = solution.group(0)
+            final_answer = (
+                final_answer.split("#### ")[1].replace(",", "").replace("$", "")
+            )
+    elif method == "flexible":
+        answer = re.findall("(\\-?[0-9\\.\\,]+)", solution_str)
+        final_answer: Optional[str] = None
+        if len(answer) == 0:
+            # No reward is there is no answer.
+            pass
+        else:
+            invalid_str: list[str] = ["", "."]
+            # Find the last number that is not '.'
+            for final_answer in reversed(answer):
+                if final_answer not in invalid_str:
+                    break
+    return final_answer
+
+
+# Reward functions need to follow a [predefined signature.](https://verl.readthedocs.io/en/latest/preparation/reward_function.html)
+
+
+def compute_reward(
+    data_source: str, solution_str: str, ground_truth: str, extra_info: dict
+) -> float:
+    answer = extract_solution(solution_str=solution_str, method="strict")
+    if answer is None:
+        return 0.0
+    else:
+        if answer == ground_truth:
+            return 1.0
+        else:
+            return 0.0
+
+
+# We then define constants to pass into VERL during the training run.
+PATH_TO_REWARD_FUNCTION: str = "/root/grpo-verl.py"
+REWARD_FUNCTION_NAME: str = "compute_reward"
+
+# ## Kicking off a training run
+
+# We define some more constants for the training run.
+MODELS_PATH: Path = Path("/models")
+MICROBATCH_SIZE_PER_GPU: int = 16
+MINUTES: int = 60
+HOURS: int = 60 * MINUTES
+DAY: int = 24 * HOURS
+
+# VERL uses Ray under the hood, and creates Ray workers, where each worker can be thought of as a python process, for each of the steps in the RL dataflow pipeline.
+# Examples of steps here include rollouts, actor forward passes, critic forward passes, etc. VERL also keeps a separate control flow process that's independent of
+# this, responsible for figuring out what step in the RL pipeline to execute. More details [here](https://verl.readthedocs.io/en/latest/hybrid_flow.html).
+
+# Each Ray worker gets mapped onto 1 or more GPUs. Depending on the number of GPUs available, Ray will decide what workers go where, or to hold off scheduling workers
+# if there are no available GPUs. Generally, more VRAM = less hot-swapping of Ray workers, which means less waiting around for memory copying each iteration.
+# In this example we have chosen a configuration that allows for easy automated testing, but you may wish to use more GPUs or more powerful GPU types.
+
+GPU_TYPE: str = "H100!"
```

Comment:
I get that we use the same configuration in both inference and training, but isn't that just a coincidence? In that case it would be better to hard code. Also let's use `H100` instead of `H100!` when we can in docs.